File: python/swarmalators/Swarmalator.py

import math
import numpy as np
import random
import colorsys
import MusicalSwarmalator

class Swarmalator:
    def __init__(self, id, isInteractive):
        self.ID = id
        self.J = 1.0#0.1
        self.K = -0.75#1.0
        self.Position = np.array([0.0,0.0,0.0])
        self.Velocity = np.array([0.0,0.0,0.0])
        self.PositionScale = 10.0
        self.Speed = 4.0
        self.Phase = 0.0
        self.Amplitude = 1.0
        self.Frequency = 1.0
        self.HueColor = 0.0
        self.oscillator = 0
        self.Others = []
        self.isInteractive = isInteractive
        if not isInteractive:
            self.MusicalAgent = MusicalSwarmalator.MusicalSwarmalator(self.ID)

    def SetOthers(self, others):
        self.Others= others

    def AssignRandomPosition(self, radius):
        self.Position = np.array([random.uniform(-radius, radius), random.uniform(-radius, radius), random.uniform(-radius, radius)])

    def ScaledPosition(self):
        return self.Position * self.PositionScale
    
    def ScaledVelocity(self):
        return self.Velocity * self.PositionScale
    
    def SetFromScalePosition(self, x, y, z):
        self.Position = np.array([x, y, z]) / self.PositionScale

    #reset oscillator by adjusting phase to start from zero
    def ResetOscillator(self, current_time):
        self.Phase = 2 * math.pi - ((2.0*math.pi * self.Frequency * current_time) % (2.0*math.pi))        


    #Time is in seconds
    def Update(self, dt, current_time):
        if not self.isInteractive:        
            (Xi, Oi) = self._delta_calculations()     
            self.Velocity = Xi * self.Speed
            self.Position += self.Velocity * dt            
            self.Phase += Oi * self.Speed * dt
            self.Phase = self.Phase % (2 * math.pi)
            self.Phase = self.Phase + (2 * math.pi) if self.Phase < 0 else self.Phase
        self.HueColor = (self.Phase / (2 * math.pi))
        self.oscillator = self.Amplitude * math.sin(2 * math.pi * self.Frequency * current_time + self.Phase)

    def _delta_calculations(self):
        Xi = np.array([0.0,0.0,0.0])
        Oi = 0
        for j in range(len(self.Others)):
            Xji = self.Others[j].Position - self.Position
            Oji = self.Others[j].Phase - self.Phase + random.uniform(0, 0.0001)
            Xji_magnitude = np.linalg.norm(Xji)            
            Xi += (Xji / Xji_magnitude) * (1 + self.J * math.cos(Oji)) - Xji / math.pow(Xji_magnitude, 3)
            Oi += math.sin(Oji) / Xji_magnitude
        N = len(self.Others) + 1  # +1 because agents does not contain "this" agent
        Xi /= N
        Oi *= (self.K / N)
        return (Xi, Oi)
    
    def CurrentColorHex(self):
        return Swarmalator.calculate_color_hex(self.HueColor, self.oscillator, self.Amplitude)
    
    #This logic is duplicated from the Update method and _delta_calculations method
    #TODO: Refactor to avoid code duplication if performance is not affected
    @staticmethod
    def calculate_params(dt, current_time, myPhase, myPosition, othersPhases, othersPositions, J, K, speed, amplitude, frequency):
        # Delta calculations
        Xi = np.zeros(3)
        Oi = 0
        num_others = len(othersPhases)  # it is assumed that othersPhases and othersPositions have the same length
        for j in range(num_others):
            Xji = othersPositions[j] - myPosition
            Oji = othersPhases[j] - myPhase + random.uniform(0, 0.0001)
            Xji_magnitude = np.linalg.norm(Xji)
            Xi += (Xji / Xji_magnitude) * (1 + J * math.cos(Oji)) - Xji / (Xji_magnitude ** 3)
            Oi += math.sin(Oji) / Xji_magnitude
        N = num_others + 1  # +1 because agents does not contain "this" agent
        Xi /= N
        Oi *= (K / N)

        #Parameter calculations
        pi2 = 2 * math.pi
        myNewVelocity = Xi * speed
        myNewPosition = myPosition + myNewVelocity * dt
        myNewPhase = (myPhase + Oi * speed * dt) % pi2
        myNewPhase = myNewPhase + pi2 if myNewPhase < 0 else myNewPhase
        hueColor = myNewPhase / pi2
        oscillator = amplitude * math.sin(pi2 * frequency * current_time + myNewPhase)

        return [myNewPosition, myNewVelocity, myNewPhase, hueColor, oscillator, Swarmalator.calculate_color_hex(hueColor, oscillator, amplitude)]
    
    @staticmethod
    def calculate_color_hex(hueColor, oscillator, amplitude):
        #Update color
        target_color = colorsys.hsv_to_rgb(hueColor, 1, 1)
        neutral_color = (0.0, 0.0, 0.0)  # black
        current_color = tuple(neutral_color[i] + (target_color[i] - neutral_color[i]) * (oscillator + amplitude) / (2 * amplitude) for i in range(3))
        return '%02x%02x%02x' % tuple(int(c * 255) for c in current_color)

File: c++/swarmalators/Swarmalator.cpp
#define _USE_MATH_DEFINES

#include "Utils.h"
#include "Swarmalator.h"
#include "MusicalSwarmalator.h"
#include <cmath>
#include <random>
#include <algorithm>
#include <sstream>
#include <iomanip>
#include <iostream>

namespace {
    std::random_device rd;
    std::mt19937 gen(rd());
}

Swarmalator::Swarmalator(int id, bool isInteractive) :
    ID(id), J(1.0), K(-0.75), Position{ 0.0, 0.0, 0.0 }, PositionScale(10.0), Speed(4.0), Velocity{ 0.0, 0.0, 0.0 },
    Phase(0.0), Amplitude(1.0), Frequency(1.0), HueColor(0.0), oscillator(0.0), IsInteractive(isInteractive),MusicalSwarmalator(id)  {}

void Swarmalator::SetOthers(const std::vector<Swarmalator*>& others) {
    Others = others;
}

void Swarmalator::AssignRandomPosition(float radius) {
    std::uniform_real_distribution<> dis(-radius, radius);
    for (float& coord : Position) {
        coord = dis(gen);
    }
}

std::array<float, 3> Swarmalator::ScaledPosition() const {
    return { Position[0] * PositionScale, Position[1] * PositionScale, Position[2] * PositionScale };
}

std::array<float, 3> Swarmalator::ScaledVelocity() const {
	return { Velocity[0] * PositionScale, Velocity[1] * PositionScale, Velocity[2] * PositionScale };
}

void Swarmalator::SetFromScaledPosition(float x, float  y, float z) {
    Position[0] = x / PositionScale;
	Position[1] = y / PositionScale;
	Position[2] = z / PositionScale;
}

//reset oscillator by adjusting phase to start from zero
void Swarmalator::ResetOscillator(float current_time) {
	Phase = 2.0 * M_PI - std::fmod(2.0f * M_PI * Frequency * current_time, 2.0f * M_PI);
}

void Swarmalator::Update(float dt, float current_time) {
    if (!IsInteractive) {
        auto result = _delta_calculations();
        auto Xi = result.first;
        auto Oi = result.second;
        for (int i = 0; i < 3; ++i) {
            Velocity[i] = Xi[i] * Speed;
            Position[i] += Velocity[i] * dt;
        }
        Phase += Oi * Speed * dt;
        Phase = std::fmod(Phase, 2.0f * M_PI);
        if (Phase < 0) Phase += 2.0f * M_PI;
    }
    HueColor = Phase / (2.0f * M_PI);
    oscillator = Amplitude * std::sin(2 * M_PI * Frequency * current_time + Phase);
}

std::pair<std::array<float, 3>, float> Swarmalator::_delta_calculations() const {
    std::array<float, 3> Xi = { 0.0, 0.0, 0.0 };
    float Oi = 0.0;
    for (auto other : Others) {
        std::array<float, 3> Xji;
        for (int i = 0; i < 3; ++i) {
            Xji[i] = other->Position[i] - Position[i];
        }
        float Oji = other->Phase - Phase + std::uniform_real_distribution<>(0.0, 0.0001)(gen);
        float Xji_magnitude = vectorMagnitude(Xji[0], Xji[1], Xji[2]);
        for (int i = 0; i < 3; ++i) {
            Xi[i] += (Xji[i] / Xji_magnitude) * (1 + J * std::cos(Oji)) - (Xji[i] / std::pow(Xji_magnitude, 3));
        }
        Oi += std::sin(Oji) / Xji_magnitude;
    }
    int N = Others.size() + 1;
    for (float& xi : Xi) {
        xi /= N;
    }
    Oi *= (K / N);
    return { Xi, Oi };
}

std::string Swarmalator::CurrentColorHex() const {
    return calculate_color_hex(HueColor, oscillator, Amplitude);
}

std::array<float, 5> Swarmalator::calculate_params(float dt, float current_time, float myPhase, const std::array<float, 3>& myPosition,
    const std::vector<float>& othersPhases, const std::vector<std::array<float, 3>>& othersPositions,
    float J, float K, float speed, float amplitude, float frequency) {
    std::array<float, 3> Xi = { 0.0, 0.0, 0.0 };
    float Oi = 0.0;
    int num_others = othersPhases.size();
    for (int j = 0; j < num_others; ++j) {
        std::array<float, 3> Xji;
        for (int i = 0; i < 3; ++i) {
            Xji[i] = othersPositions[j][i] - myPosition[i];
        }
        float Oji = othersPhases[j] - myPhase + std::uniform_real_distribution<>(0.0, 0.0001)(gen);
        float Xji_magnitude = vectorMagnitude(Xji[0], Xji[1], Xji[2]);
        for (int i = 0; i < 3; ++i) {
            Xi[i] += (Xji[i] / Xji_magnitude) * (1 + J * std::cos(Oji)) - (Xji[i] / std::pow(Xji_magnitude, 3));
        }
        Oi += std::sin(Oji) / Xji_magnitude;
    }
    int N = num_others + 1;
    for (float& xi : Xi) {
        xi /= N;
    }
    Oi *= (K / N);

    float pi2 = 2 * M_PI;
    std::array<float, 3> myNewPosition;
    for (int i = 0; i < 3; ++i) {
        myNewPosition[i] = myPosition[i] + Xi[i] * speed * dt;
    }
    float myNewPhase = std::fmod(myPhase + Oi * speed * dt, pi2);
    if (myNewPhase < 0) myNewPhase += pi2;
    float hueColor = myNewPhase / pi2;
    float oscillator = amplitude * std::sin(pi2 * frequency * current_time + myNewPhase);

    return { myNewPosition[0], myNewPosition[1], myNewPosition[2], hueColor, oscillator };
}

std::string Swarmalator::calculate_color_hex(float hueColor, float oscillator, float amplitude) {
    auto hsv_to_rgb = [](float h, float s, float v) {
        float r, g, b;
        int i = static_cast<int>(h * 6);
        float f = h * 6 - i;
        float p = v * (1 - s);
        float q = v * (1 - f * s);
        float t = v * (1 - (1 - f) * s);

        switch (i % 6) {
        case 0: r = v, g = t, b = p; break;
        case 1: r = q, g = v, b = p; break;
        case 2: r = p, g = v, b = t; break;
        case 3: r = p, g = q, b = v; break;
        case 4: r = t, g = p, b = v; break;
        case 5: r = v, g = p, b = q; break;
        }
        return std::array<float, 3> {r, g, b};
        };

    std::array<float, 3> target_color = hsv_to_rgb(hueColor, 1.0f, 1.0f);
    std::array<float, 3> neutral_color = { 0.0f, 0.0f, 0.0f };
    std::array<float, 3> current_color;
    for (int i = 0; i < 3; ++i) {
        current_color[i] = neutral_color[i] + (target_color[i] - neutral_color[i]) * (oscillator + amplitude) / (2 * amplitude);
    }

    std::ostringstream oss;
    oss << std::hex << std::setfill('0') << std::setw(2) << static_cast<int>(current_color[0] * 255)
        << std::setw(2) << static_cast<int>(current_color[1] * 255)
        << std::setw(2) << static_cast<int>(current_color[2] * 255);
    return oss.str();
}
